---
title: _signatureSearchData_ - Annotation data package for signatureSearch
author: "Author: Yuzhu Duan (yduan004@ucr.edu)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 4
    fig_caption: yes
    code_folding: show
    number_sections: true
  pdf_document:
    toc: true
fontsize: 15pt
vignette: >
  %\VignetteIndexEntry{signatureSearchData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>

```{r setup0, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This is an annotation data package for `signatureSearch` softeware packge. It
stores the CMAP/LINCS reference databases used for GESS methods in the
`signatureSearch` package as well as other intermediate datasets.

# `cmap` signature database
The `cmap` database represents the $log_2$ fold change (log2FC) of 12,437 genes 
from 1,281 compound treatments in 4 cells (3,478 signatures in total). 
It is stored as HDF5 backed `SummarizedExperiment` object and can be loaded 
through the `AnnotationHub` interface
```{r cmap, eval=FALSE}
library(AnnotationHub)
ah <- AnnotationHub()
qr <- query(ah, c("signatureSearchData", "cmap"))
qr
cmap <- load_sigdb("AH69075","AH69076","cmap")
```

## Generate the `cmap` database from the original resources
The resource data was downloaded from the 
[Connectivity Map](https://portals.broadinstitute.org/cmap/) project, 
which is the CEL files generated by Affymetrix chips. 
The downloaded CEL files were normalized by `MAS5` method. 
Then the `limma` package was used to do the differential expression (DE)
analysis to get the `logMA` matrix containing the log2 fold change scores. 
The `logMA` matrix was used to build the HDF5 backed `SummarizedExperiment` 
object, which is the `cmap` database.
The following steps were used to generate the `cmap` database.

### Create expected directory structure
The following creates the expected directory structure. Input data will be 
stored in the data directory and results will be written to the results 
directory. All paths are given relative to the present working directory of 
the userâ€™s R session.
```{r work_envir, eval=FALSE}
dir.create("data"); dir.create("data/CEL"); dir.create("results") 
```

### Download data from Connectivity Map project site
The drug-related expression data are downloaded from the CMAP web site
[here](http://www.broadinstitute.org/cmap). The `getCmapCEL` function
downloads the 7,056 CEL files. The functions will write the
downloaded files to the `data/CEL` directories within the present 
working directory of the user's R session. Since some of the raw data sets 
are large, the functions will only rerun the download if the argument `rerun` 
is assigned `TRUE`. If the raw data are not needed then users can skip this 
time consuming download step and work with the preprocessed data 
obtained in the next section.

```{r download_cmap, eval=FALSE}
getCmapCEL(rerun=FALSE) # Download cmap CEL files. this will take some time
```

The experimental design of the CMAP project is defined in the file
`cmap_instances_02.xls`.  Note, this file required some cleaning in LibreOffice
(Excel would work for this too). After this it was saved as tab delimited txt
file named `cmap_instances_02.txt`. It is stored in this data package.
```{r cmap_inst2, eval=TRUE}
path <- system.file("extdata", "cmap_instances_02.txt", package="signatureSearchData")
cmap_inst <- read.delim(path, check.names=FALSE) 
```

The panel of cell lines used by CMAP includes 
[MCF7](http://www.broadinstitute.org/cmap/help_topics_linkified.jsp#mcf7), 
[ssMCF7](http://www.broadinstitute.org/cmap/help_topics_linkified.jsp#mcf7), 
[HL60](http://www.broadinstitute.org/cmap/help_topics_linkified.jsp#hl60), 
[PC3](http://www.broadinstitute.org/cmap/help_topics_linkified.jsp#pc3) and 
[SKMEL5](http://www.broadinstitute.org/cmap/help_topics_linkified.jsp#skmel5). 

### Determine chip type from CEL files 
The CMAP data set is based on three different Affymetrix chip types (HG-U133A,
HT_HG-U133A and U133AAofAv2). The following extracts the chip type information
from the CEL files and stores the result in an `rds` file with the path 
`./data/chiptype.rds`. Users who skipped the download of the CEL files can
download this file [here](http://biocluster.ucr.edu/~tgirke/projects/longevity/cmap/data/chiptype.rds).

```{r get_cel_type, eval=FALSE}
library(affxparser)
celfiles <- list.files("./data/CEL", pattern=".CEL$")
chiptype <- sapply(celfiles, function(x) affxparser::readCelHeader(paste0("data/CEL/", x))$chiptype)
if(FALSE) saveRDS(chiptype, "./data/chiptype.rds") # if(FALSE) protects this line from accidental execution!
```

### Normalization of CEL files

The follwoing processes the CEL files from each chip type separately using the
MAS5 normalization algorithm. The results will be written to 3 subdirectores
under `data` that are named after the chip type names. To save time, the
processing is parallelized with `BiocParallel` to run on 100 CPU cores of a
computer cluster with a scheduler (_e.g._ Torque). The number of CEL files from
each chip type are: 807 CEL files from HG-U133A, 6029 CEL files from
HT_HG-U133A, and 220 CEL files from U133AAofAv2. Note, these numbers are slightly
different than those reported in the `cmap_instances_02.txt` file. 
The MAS5 normalized data sets can be downloaded here: 
[HG-U133A](http://biocluster.ucr.edu/~tgirke/projects/longevity/cmap/data/HG-U133A/all_mas5exprs.rds), 
[HT_HG-U133A](http://biocluster.ucr.edu/~tgirke/projects/longevity/cmap/data/HT_HG-U133A/all_mas5exprs.rds), 
[U133AAofAv2](http://biocluster.ucr.edu/~tgirke/projects/longevity/cmap/data/U133AAofAv2/all_mas5exprs.rds).

```{r normalize_chips, eval=FALSE}
chiptype <- readRDS("./data/chiptype.rds")
chiptype_list <- split(names(chiptype), as.character(chiptype))
normalizeCel(chiptype_list, rerun=FALSE)
```

Combine results from same chip type in single data frame
```{r comb_chip_type_data, eval=FALSE}
chiptype_dir <- unique(readRDS("./data/chiptype.rds"))
combineResults(chiptype_dir, rerun=FALSE)
```

### Clean-up of intermediate files

This deletes intermediate files. Before executing these lines, 
please make sure that this is what you want.

```{r cleanup1, eval=FALSE}
for(i in seq_along(chiptype_dir)) 
    unlink(list.files(paste0("data/", chiptype_dir[i]), pattern="cellbatch", 
                      full.names=TRUE), recursive=TRUE)
unlink("data/CEL/*.CEL") # Deletes downloaded CEL files
```

### Obtain annotation information
The following generates annotation information for the Affymetirx probe set
identifiers. Note, the three different Affymetrix chip types used by CMAP
share most probe set ids (>95%), meaning it is possible to combine the data
after normalization and use the same annotation package for all of them. The
annotation libraries for the chip types HG-U133A and HT_HG-U133A are
`hgu133a.db` and `hthgu133a.db`, respectively. However, there is no annotation 
library (e.g. CDF) available for U133AAofAv2. The annotation file can be downloaded
from here: [`myAnnot.xls`](http://biocluster.ucr.edu/~tgirke/projects/longevity/cmap/results/myAnnot.xls).

```{r affyid_annotations, eval=FALSE, message=FALSE}
library(hgu133a.db)
myAnnot <- data.frame(ACCNUM=sapply(contents(hgu133aACCNUM), paste, collapse=", "), 
                             SYMBOL=sapply(contents(hgu133aSYMBOL), paste, collapse=", "), 
                             UNIGENE=sapply(contents(hgu133aUNIGENE), paste, collapse=", "), 
                             ENTREZID=sapply(contents(hgu133aENTREZID), paste, collapse=", "), 
                             ENSEMBL=sapply(contents(hgu133aENSEMBL), paste, collapse=", "), 
                             DESC=sapply(contents(hgu133aGENENAME), paste, collapse=", "))
write.table(myAnnot, file="./results/myAnnot.xls", quote=FALSE, sep="\t", col.names = NA) 
saveRDS(myAnnot, "./results/myAnnot.rds")
```

### DEG analysis with `limma`

#### Generate list of CEL names defining treatment vs. control comparisons

The `sampleList` function extracts the sample comparisons (contrasts) from the
CMAP annotation table and stores them as a list.

```{r cel_file_list, eval=FALSE}
path <- system.file("extdata", "cmap_instances_02.txt", package="signatureSearchData")
cmap_inst <- read.delim(path, check.names=FALSE) 
comp_list <- sampleList(cmap_inst, myby="CMP_CELL")
```

#### Load normalized expression data 

The following loads the MAS5 normalized expression data into a single `data.frame`. 
To accelerate the import, the data is read from `rds` files. 

```{r load_mas5_data, eval=FALSE}
chiptype_dir <- unique(readRDS("./data/chiptype.rds"))
df1 <- readRDS(paste0("data/", chiptype_dir[1], "/", "all_mas5exprs.rds"))
df2 <- readRDS(paste0("data/", chiptype_dir[2], "/", "all_mas5exprs.rds"))
df3 <- readRDS(paste0("data/", chiptype_dir[3], "/", "all_mas5exprs.rds"))
affyid <- rownames(df1)[rownames(df1) %in% rownames(df2)]; affyid <- affyid[affyid %in% rownames(df3)]
mas5df <- cbind(df1[affyid,], df2[affyid,], df3[affyid,])
```

### Transform probe set to gene level data
The next step generates gene level expression values. If genes are represented by several
probe sets then their mean intensities are used. This is necessary because 
the U133 chip contains many genes with duplicated probe sets. Probe sets not matching 
any gene are removed.

```{r mean_mas5_data, eval=FALSE} 
myAnnot <- readRDS("./results/myAnnot.rds") 
myAnnot <- myAnnot[as.character(myAnnot[,"ENTREZID"]) != "NA",]
mas5df <- mas5df[rownames(myAnnot),]
idlist <- tapply(row.names(myAnnot), as.character(myAnnot$ENTREZID), c)
mas5df <- t(sapply(names(idlist), function(x) colMeans(mas5df[idlist[[x]], ])))
```

### DEG analysis with `limma`

The analysis of differentially expressed genes (DEGs) is performed with the `limma` package.
The log2 fold change (log2FC) scores of drug treatments in cells are used as 
gene expression profiles (GEPs). The log2FC matrix (`logMA`) is stored in the generated [`degList`](http://biocluster.ucr.edu/~yduan004/signatureSearchData/degList.rds).

```{r deg_limma, eval=FALSE}
degList <- runLimma(df=log2(mas5df), comp_list, fdr=0.10, foldchange=1, verbose=TRUE)
saveRDS(degList, "./results/degList.rds") # saves entire degList
```

### Convert the `logMA` into HDF5 backed `SummarizedExperiment` object  
The generated HDF5 backed `SummarizedExperiment` object is the `cmap` database 
storing the log2FC scores of compound treatments in cells.
```{r se, eval=FALSE}
library(SummarizedExperiment)
library(HDF5Array)
degList <- readRDS("./results/degList.rds")
logMA <- degList$logFC
pert_cell_factor = gsub("(^.*)_(.*$)", "\\1__\\2__trt_cp", colnames(logMA))
## Add colData annotation for SummarizedExperiment object
colData <- DataFrame(pert_cell_factor=pert_cell_factor,
                     pert_cell = colnames(logMA))
new <- as.data.frame(t(sapply(seq_len(nrow(colData)), function(i) 
  unlist(strsplit(as.character(colData$pert_cell[i]), "_")))), 
  stringsAsFactors=FALSE)
colnames(new) = c("pert", "cell")
colData <- cbind(colData, as(new, "DataFrame"))
rownames(colData) = colData$pert_cell_factor
colnames(logMA)=colData$pert_cell_factor # Important to match the treatment format
cmap <- SummarizedExperiment(assays = list(score=logMA), colData = colData)
## Save 'cmap' as an HDF5-based SummarizedExperiment object on disk. 
## Here, it will save to the data directory of the userâ€™s R session.
cmap <- saveHDF5SummarizedExperiment(cmap, "./data/cmap")
```

# `cmap_expr` signature database
For correlation based GESS methods (`gess_cor`) in the `signatureSearch` software 
package, the signature database containing gene expression intensity values of drug 
treatment samples could also be used to search for similarity when the query
signature are also gene expression values. 

The `cmap_expr` database represents mean expression values of 1,309 drug
treatment samples in 4 cells (3,587 signatures in total).
It is stored as HDF5 backed `SummarizedExperiment` object and can be loaded
through the `AnnotationHub` interface
```{r cmap_expr, eval=FALSE}
library(AnnotationHub)
ah <- AnnotationHub()
qr <- query(ah, c("signatureSearchData", "cmap"))
cmap_expr <- load_sigdb("AH69077","AH69078","cmap_expr")
```

## Generate the `cmap_expr` database from the original resources
The same as the `cmap` database, the resource data was downloaded from the 
[Connectivity Map](https://portals.broadinstitute.org/cmap/) project. 
The gene expression intensity data was rma normalized. Then, the mean expression 
values of drug treatment samples at different concentration, duration was used
as expression values of that drug in a specific cell. The expression matrix
was saved as HDF5-based SummarizedExperiment object (cmap_expr database).

### Normalization of CEL files with rma method
Use the same work environment as preparing the `cmap` database.
```{r norm_rma, eval=FALSE}
chiptype <- readRDS("./data/chiptype.rds") 
chiptype_list <- split(names(chiptype), as.character(chiptype))
## need to allocate large number of memory, 512Gb works
normalizeCel_rma(chiptype_list, rerun=FALSE) 
df1 <- read.delim("./data/HG-U133A/rma_exprs.xls", sep="\t", header=TRUE, 
                  row.names=1, check.names=FALSE)
df2 <- read.delim("./data/HT_HG-U133A/rma_exprs.xls", sep="\t", header=TRUE, 
                  row.names=1, check.names=FALSE)
df3 <- read.delim("./data/U133AAofAv2/rma_exprs.xls", sep="\t", header=TRUE, 
                  row.names=1, check.names=FALSE)
affyid <- rownames(df1)[rownames(df1) %in% rownames(df2)]
affyid <- affyid[affyid %in% rownames(df3)]
rmadf <- cbind(df1[affyid,], df2[affyid,], df3[affyid,])
```

### Transform probe set to gene level data
```{r prof2gene, eval=FALSE}
myAnnot <- readRDS("./results/myAnnot.rds") 
myAnnot <- myAnnot[as.character(myAnnot[,"ENTREZID"]) != "NA",]
rmadf <- rmadf[rownames(myAnnot),]
idlist <- tapply(row.names(myAnnot), as.character(myAnnot$ENTREZID), c)
rmadf <- t(sapply(names(idlist), function(x) colMeans(rmadf[idlist[[x]], ])))
saveRDS(rmadf,"./data/rmadf.rds")
```

### Take mean expression values of drug treatment sample

Take mean expression values of multiple drug treatment samples at different 
concentration and duration in the same cell as expression value of that drug
treatment in the cell.
```{r rma2cmap_expr, eval=FALSE}
rmadf <- readRDS("./data/rmadf.rds") # matrix: 12437 7056
path <- system.file("extdata", "cmap_instances_02.txt", package="signatureSearchData")
cmap_inst <- read.delim(path, check.names=FALSE) 
cmap_inst <- data.frame(cmap_inst, drug_cell = paste(cmap_inst$cmap_name,  
                                                     cmap_inst$cell2, sep ="_"))
celid <- paste0(cmap_inst$perturbation_scan_id,".CEL")
sum(celid %in% colnames(rmadf)) # 5408
celid2 <- gsub("'", "", celid) # get rid of "'" in celid
sum(celid2 %in% colnames(rmadf)) # 6100
drug_celid_list <- tapply(celid2, cmap_inst$drug_cell, c) # 3587
cmap_drug_cell_expr <- sapply(names(drug_celid_list), function(x) 
  rowMeans(as.data.frame(rmadf[,drug_celid_list[[x]]]))) # 12437 X 3587
saveRDS(cmap_drug_cell_expr, "./data/cmap_drug_cell_expr.rds")
```

### Generate `cmap_expr` database from the expression matrix
```{r gen_cmap_expr, eval=FALSE}
cmap_drug_cell_expr <- readRDS("./data/cmap_drug_cell_expr.rds")
## match format of colnames of "cmap_drug_cell_expr" as "pert__cell__trt_cp" 
pert_cell_factor = gsub("(^.*)_(.*$)", "\\1__\\2__trt_cp", 
                        colnames(cmap_drug_cell_expr))
colnames(cmap_drug_cell_expr) <- pert_cell_factor
cmap_expr <- SummarizedExperiment(assays = list(expr=cmap_drug_cell_expr))
## Save 'cmap_expr' as an HDF5-based SummarizedExperiment object:
cmap_expr <- saveHDF5SummarizedExperiment(cmap_expr, "./data/cmap_expr")
```

# `lincs` signature database
The `lincs` database represents moderated z-scores from DE analysis of 12,328 
genes from 8,140 compound treatments in 30 cells (45,956 signatures in total). 
It can be loaded through the `AnnotationHub` interface

```{r lincs, eval=FALSE}
library(AnnotationHub)
ah <- AnnotationHub()
qr <- query(ah, c("signatureSearchData", "lincs"))
lincs <- load_sigdb("AH69079","AH69080","lincs")
```

## Generate the `lincs` database from the original resources

The original LINCS Level 5 moderated z-score database was downloaded at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742). 
The downloaded database in `gctx` format was processed and saved as HDF5 backed
`SummarizedExperiment` object 

### Download the original LINCS Level5 datasets at GEO 

Download the following files from [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742):

+ GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx.gz
+ GSE92742_Broad_LINCS_gene_info.txt.gz
+ GSE92742_Broad_LINCS_sig_info.txt.gz
+ GSE92742_Broad_LINCS_cell_info.txt.gz

Unzip the downloaded files to the `data` direcotry under the present working 
directory of the userâ€™s R session.

The LINCS Level5 signatures represent moderated z-scores of compound treatments
at different contration and time in different cells from differential expression
analysis, the replicates have been weighted averaged.

### Filter samples at specific concentation and treatment time
```{r filter_meta42, eval=FALSE}
library(readr); library(dplyr); library(magrittr);
library(SummarizedExperiment); library(HDF5Array)
meta42 <- read_tsv("./data/GSE92742_Broad_LINCS_sig_info.txt") 
txt <- names(table(meta42$pert_idose))
unit <- txt[grep("10 .M", txt)][2]
meta42 %<>% filter(pert_type=="trt_cp" & pert_idose==unit & pert_itime=="24 h") 
# filter rows by compound, 10 uM concentration, and 24 h treatment time
meta42 %<>% 
  bind_cols(alt_id=paste(meta42$pert_iname, meta42$cell_id, sep="_")) %>%
  bind_cols(pert_cell_factor=paste(meta42$pert_iname, meta42$cell_id, 
                                   meta42$pert_type, sep="__")) %>% 
  distinct(alt_id, .keep_all=TRUE) # eliminate technical duplicates
dim(meta42) # 45956 X 14
```

### Extract z-scores matrix from gctx file 

Extract z-scores of selected samples from gctx file.
The following codes require large number of memory in the system, 50Gb works
```{r extract_modz, eval=FALSE}
lincs42_mat <- parse_gctx("./data/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx", 
                          cid=meta42$sig_id, matrix_only=TRUE)
lincs42_mat <- lincs42_mat@mat
colnames(lincs42_mat) = meta42$pert_cell_factor 
```

### Convert the z-score matrix to HDF5 backed SummarizedExperiment object
```{r gen_lincs, eval=FALSE}
## Add rowData annotation
geneInfo <- read_tsv("./data/GSE92742_Broad_LINCS_gene_info.txt")
geneInfo <- as(geneInfo, "DataFrame")
geneInfo$pr_gene_id <- as.character(geneInfo$pr_gene_id)
rownames(geneInfo) <- geneInfo$pr_gene_id
geneInfo <- geneInfo[rownames(lincs42_mat),]
meta42 <- as(meta42, "DataFrame")
rownames(meta42) <- meta42$pert_cell_factor
lincs42 <- SummarizedExperiment(assays = list(score = lincs42_mat), 
                                rowData = geneInfo, colData = meta42)
## Add cell type information to metadata
cell_info_path <- system.file("extdata", "cell_info.tsv", package="signatureSearchData")
cell_info <- read_tsv(cell_info_path)
## The cell_info is from the "GSE92742_Broad_LINCS_cell_info.txt" with some edition
metadata(lincs42)$cell_info = cell_info
# Save 'lincs42' as an HDF5-based SummarizedExperiment object:
lincs42 <- saveHDF5SummarizedExperiment(lincs42, "./data/lincs")
```

# `lincs_expr` signature database
The `lincs_expr` database contains mean gene expression values 
from 5,925 compound treatment samples in 30 cells (38,824 signatures in total). 
It can be loaded through the `AnnotationHub` interface
```{r lincs_expr, eval=FALSE}
library(AnnotationHub)
ah <- AnnotationHub()
qr <- query(ah, c("signatureSearchData", "lincs"))
lincs_expr <- load_sigdb("AH69081","AH69082","lincs_expr")
```

## Generate the `lincs_expr` database from the original resources
The original LINCS Level3 expression database was downloaded at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742). 
The downloaded database in `gctx` format was processed and saved as HDF5 backed
`SummarizedExperiment` object

### Download the original LINCS Level3 datasets at GEO 

Download the following files from [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742):

+ GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx.gz
+ GSE92742_Broad_LINCS_gene_info.txt.gz
+ GSE92742_Broad_LINCS_inst_info.txt.gz

Unzip the downloaded files to the `data` direcotry under the present working 
directory of the userâ€™s R session.

The LINCS Level3 signatures represent normalized expression intensity values of 
compound treatment samples at different contration and time in different cells, 
including replicates.

### Filter samples at specific concentation and treatment time
```{r filter_expr, eval=FALSE}
inst42 <- read_tsv("./data/GSE92742_Broad_LINCS_inst_info.txt") 
inst42 %<>% filter(pert_type=="trt_cp" & pert_dose==10 & pert_dose_unit=="um" 
                   & pert_time==24 & pert_time_unit=="h")
inst42 %<>% bind_cols(alt_id=paste(inst42$pert_iname, inst42$cell_id, sep="_")) %>% 
  bind_cols(pert_cell_factor=paste(inst42$pert_iname, inst42$cell_id, inst42$pert_type, sep="__"))
dim(inst42) # 169,795 X 13
```

### Extract expression matrix from gctx file 

Extract expression values of selected samples from gctx file.
The following codes require large number of memory in the system, 50Gb works
```{r extract_expr, eval=FALSE}
lincs42_mat3 <-parse_gctx("./data/GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx", 
                          cid=inst42$inst_id, matrix_only=TRUE)
lincs42_mat3 <- lincs42_mat3@mat # 12328 X 169795
## Get mean expression values of samples from the same drug treatment in cells
pert_cell_list <- split(inst42$inst_id, inst42$pert_cell_factor)
lincs_drug_cell_expr <- sapply(names(pert_cell_list), function(x) 
  rowMeans(as.data.frame(lincs42_mat3[,pert_cell_list[[x]]]))) # 12328 X 38824
saveRDS(lincs_drug_cell_expr, "./data/lincs_drug_cell_expr.rds")
```

### Convert the expression matrix to HDF5 backed SummarizedExperiment object
```{r expr2se, eval=FALSE}
## Add colData annotation
inst42_uniq <- inst42[!duplicated(inst42$pert_cell_factor),] # 38824 X 13
inst42_uniq <- as(inst42_uniq, "DataFrame")
rownames(inst42_uniq) <- inst42_uniq$pert_cell_factor
lincs42_expr <- SummarizedExperiment(
  assays = list(expr = lincs_drug_cell_expr), 
  rowData = geneInfo, 
  colData = inst42_uniq[colnames(lincs_drug_cell_expr),])
## Save 'lincs42_expr' as an HDF5-based SummarizedExperiment object:
lincs42_expr <- saveHDF5SummarizedExperiment(lincs42_expr, "./data/lincs_expr")
```

# Other datasets
This annotation package also contains other intermediate datasets used for
functionalities in the `signatureSearch` software package, 
such as getting targets of query drugs in [DrugBank](https://www.drugbank.ca/), 
[CLUE](https://clue.io/) and [STITCH](http://stitch.embl.de/) databases, 
conducting GO enrichment analysis or computing p values or Tau scores of the
`gess_lincs` method.

