---
title: _signatureSearch_data_ - Annotation data package for signatureSearch
author: "Author: Yuzhu Duan (yduan004@ucr.edu)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  html_document:
    toc: true
    toc_float:
        collapsed: true
        smooth_scroll: true
    toc_depth: 4
    fig_caption: yes
    code_folding: show
    number_sections: true
  pdf_document:
    toc: true
fontsize: 15pt
vignette: >
  %\VignetteIndexEntry{signatureSearch_data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>

```{r setup0, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This is an annotation data package for `signatureSearch` softeware packge. It
stores the CMAP/LINCS reference databases used for GESS methods in the
`signatureSearch` package as well as other intermediate datasets.

# `cmap` signature database
The `cmap` database represents the $log_2$ fold change (log2FC) of 12,437 genes 
from 1,281 compound treatments in 4 cells (3,478 signatures in total). 
It is stored as HDF5 backed `SummarizedExperiment` object and can be loaded 
through the `AnnotationHub` interface
```{r cmap, eval=FALSE}
library(AnnotationHub)
hub <- AnnotationHub()
myfiles <- query(hub, "SEARCHTERM1", "SEARCHTERM2")
myfiles[[1]]  ## load the first resource in the list
```

The resource data was downloaded from the 
[Connectivity Map](https://portals.broadinstitute.org/cmap/) project, 
which is the CEL files generated by Affymetrix chips. 
The downloaded CEL files were normalized by MAS5 method. 
Then the `limma` package was used to do the differential expression (DE)
analysis to get the `logMA` matrix containing the log2FC scores. 
The code used for DE analysis is vailable at the [Longevity Tools](http://girke.bioinformatics.ucr.edu/longevityTools/mydoc/mydoc_longevityTools_eDRUG_01.html).
The `logMA` matrix was used to build the HDF5 backed `SummarizedExperiment` 
object, which is the `cmap` database.

# `cmap_expr` signature database
For correlation based GESS methods in `signatureSearch` package, 
the signature database containing gene 
expression values from drug treatment samples could also be used to search 
for similarity. 

The `cmap_expr` database represents mean expression values of 1,309 drug
treatment samples in 4 cells (3,587 signatures in total).
It is stored as HDF5 backed `SummarizedExperiment` object and can be loaded
through the `AnnotationHub` interface
```{r cmap_expr, eval=FALSE}
library(AnnotationHub)
hub <- AnnotationHub()
myfiles <- query(hub, "SEARCHTERM1", "SEARCHTERM2")
myfiles[[1]]  ## load the first resource in the list
```

The same as the `cmap` database, the resource data was downloaded from the 
[Connectivity Map](https://portals.broadinstitute.org/cmap/) project. 
The expression intensity data was rma normalized. Then, the mean expression 
values of drug treatment samples at different concentration, duration was used
as expression values of that drug in a specific cell. The expression matrix
was saved as HDF5-based SummarizedExperiment object (cmap_expr database).

# `lincs` signature database
The `lincs` database represents moderated z-scores from DE analysis of 12,328 
genes from 8,140 compound treatments in 30 cells (45,956 signatures in total). 
The original LINCS Level 5 moderated z-score database was downloaded at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742). 
The downloaded database in `gctx` format was processed and saved as HDF5 backed
`SummarizedExperiment` object and can be loaded through the `AnnotationHub` 
interface
```{r lincs, eval=FALSE}
library(AnnotationHub)
hub <- AnnotationHub()
myfiles <- query(hub, "SEARCHTERM1", "SEARCHTERM2")
myfiles[[1]]  ## load the first resource in the list
```

# `lincs_expr` signature database
The `lincs_expr` database contains mean gene expression values 
from 5,925 compound treatment samples in 30 cells (38,824 signatures in total). 
The original LINCS Level 3 expression database was downloaded at [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742). 
The downloaded database in `gctx` format was processed and saved as HDF5 backed
`SummarizedExperiment` object and can be loaded through the `AnnotationHub` 
interface
```{r lincs_expr, eval=FALSE}
library(AnnotationHub)
hub <- AnnotationHub()
myfiles <- query(hub, "SEARCHTERM1", "SEARCHTERM2")
myfiles[[1]]  ## load the first resource in the list
```

# Other datasets
This annotation package also contains other intermediate datasets used for
functionalities in the `signatureSearch` software package, 
such as getting targets of query drugs in [DrugBank](https://www.drugbank.ca/), 
[CLUE](https://clue.io/) and [STITCH](http://stitch.embl.de/) databases, 
conducting GO enrichment analysis or computing p-values or Tau scores 
in the results from `gess_lincs` method.

For codes and details of generating the above data, please refer to the [make-data.R](https://github.com/yduan004/signatureSearch_data/blob/master/inst/scripts/make-data.R) 
script at [github](https://github.com/yduan004/signatureSearch_data).

