f#######################
## make cmap database ##
########################

# The `cmap` database represents the $log_2$ fold change of 12,437 genes 
# from 1,281 compound treatments in 4 cells (3,478 signatures in total).

# Download data from Connectivity Map project site at
# < https://portals.broadinstitute.org/cmap/>, which is the original CEL files 
# generated by Affymetrix chips. 
# The downloaded CEL files were normalized by MAS5 method and the differential 
# expression analysis from 'limma' package was used to get the `logMA` matrix 
# containing the log2FC scores. The code used for DE analysis is vailable at 
# the Longevity Tools (LT) website:
# <http://girke.bioinformatics.ucr.edu/longevityTools/mydoc/mydoc_longevityTools_eDRUG_01.html>.
# 
# The `degList.rds` generated from the above vignette is used to generate the 
# cmap log2FC database
library(SummarizedExperiment)
library(HDF5Array)
source("fct.R") # at the same place as this R script, contains some functions used below 
degList <- readRDS("degList.rds")
logMA <- degList$logFC
pert_cell_factor = gsub("(^.*)_(.*$)", "\\1__\\2__trt_cp", colnames(logMA))
colData <- DataFrame(pert_cell_factor=pert_cell_factor,
                     pert_cell = colnames(logMA))
new <- as.data.frame(t(sapply(seq_len(nrow(colData)), function(i) 
  unlist(strsplit(as.character(colData$pert_cell[i]), "_")))), 
  stringsAsFactors=FALSE)
colnames(new) = c("pert", "cell")
colData <- cbind(colData, as(new, "DataFrame"))
rownames(colData) = colData$pert_cell_factor
colnames(logMA)=colData$pert_cell_factor # Important to match format of lincs
cmap <- SummarizedExperiment(assays = list(score=logMA), colData = colData)
## Save 'cmap' as an HDF5-based SummarizedExperiment object on disk. 
## Here, it will save to the present working directory of the user’s R session.
cmap <- saveHDF5SummarizedExperiment(cmap, "./cmap")

#############################
## make cmap_expr database ##
#############################

# The `cmap_expr` database represents mean expression values of 1,309 drug
# treatment samples in 4 cells (3,587 signatures in total).

# For correlation based methods, the signature database containing gene 
# expression values from drug treatment samples could also be used to search 
# for similarity. 
# The `cmap_expr` database containing the rma normalized gene expression data.
# Then, it takes mean expression values of drug treatment samples at different 
# concentration, duration as expression values of that drug in a specific cell.

## Code used to generate the rma normalized expression dataset is similar to 
## the vignette at LT website with some modifications

### Download data from Connectivity Map project site
library(longevityTools)
getCmapCEL(rerun = FALSE) # Download cmap CEL files, will take some time
celfiles <- list.files("./data/CEL", pattern=".CEL$")
### Normalization of CEL files
chiptype <- readRDS("./data/chiptype.rds") # download `chiptype.rds` from LT.
chiptype_list <- split(names(chiptype), as.character(chiptype))
normalizeCel(chiptype_list, rerun=FALSE) # need to allocate a lot of memory, 512Gb works
df1 <- read.delim("./data/HG-U133A/rma_exprs.xls", sep="\t", header=T, 
                  row.names=1, check.names=FALSE)
df2 <- read.delim("./data/HT_HG-U133A/rma_exprs.xls", sep="\t", header=T, 
                  row.names=1, check.names=FALSE)
df3 <- read.delim("./data/U133AAofAv2/rma_exprs.xls", sep="\t", header=T, 
                  row.names=1, check.names=FALSE)
affyid <- rownames(df1)[rownames(df1) %in% rownames(df2)]
affyid <- affyid[affyid %in% rownames(df3)]
rmadf <- cbind(df1[affyid,], df2[affyid,], df3[affyid,])
### Obtain annotation information
library(hgu133a.db)
myAnnot <- data.frame(ACCNUM=sapply(contents(hgu133aACCNUM), paste, collapse=", "),
                      SYMBOL=sapply(contents(hgu133aSYMBOL), paste, collapse=", "),
                      UNIGENE=sapply(contents(hgu133aUNIGENE), paste, collapse=", "),
                      ENTREZID=sapply(contents(hgu133aENTREZID), paste, collapse=", "),
                      ENSEMBL=sapply(contents(hgu133aENSEMBL), paste, collapse=", "),
                      DESC=sapply(contents(hgu133aGENENAME), paste, collapse=", "))
write.table(myAnnot, file="./results/myAnnot.xls", quote=FALSE, sep="\t", col.names = NA)
saveRDS(myAnnot, "./results/myAnnot.rds")
### Transform probe set to gene level data
myAnnot <- readRDS("./results/myAnnot.rds") 
myAnnot <- myAnnot[as.character(myAnnot[,"ENTREZID"]) != "NA",]
rmadf <- rmadf[rownames(myAnnot),]
idlist <- tapply(row.names(myAnnot), as.character(myAnnot$ENTREZID), c)
rmadf <- t(sapply(names(idlist), function(x) colMeans(rmadf[idlist[[x]], ])))
saveRDS(rmadf,"./data/rmadf.rds")

## Generate `cmap_expr` database from `rmadf`
rmadf <- readRDS("./data/rmadf.rds") # matrix: 12437 7056
cmap_inst <- read.delim("./data/cmap_instances_02.txt", 
                        check.names=FALSE) # Download from LT website.
cmap_inst <- data.frame(cmap_inst, 
                        drug_cell = paste(cmap_inst$cmap_name, 
                                          cmap_inst$cell2, sep ="_"))
celid <- paste0(cmap_inst$perturbation_scan_id,".CEL")
sum(celid %in% colnames(rmadf)) # 5408
celid2 <- gsub("'", "", celid) # get rid of "'" in celid
sum(celid2 %in% colnames(rmadf)) # 6100
drug_celid_list <- tapply(celid2, cmap_inst$drug_cell, c) # 3587
cmap_drug_cell_expr <- sapply(names(drug_celid_list), function(x) 
  rowMeans(as.data.frame(rmadf[,drug_celid_list[[x]]]))) # 12437 X 3587
saveRDS(cmap_drug_cell_expr, "./data/cmap_drug_cell_expr.rds")
cmap_drug_cell_expr <- readRDS("./data/cmap_drug_cell_expr.rds")
## match format of colnames of "cmap_drug_cell_expr" as "pert__cell__trt_cp" 
pert_cell_factor = gsub("(^.*)_(.*$)", "\\1__\\2__trt_cp", 
                        colnames(cmap_drug_cell_expr))
colnames(cmap_drug_cell_expr) <- pert_cell_factor
cmap_expr <- SummarizedExperiment(assays = list(expr=cmap_drug_cell_expr))
## Save 'cmap_expr' as an HDF5-based SummarizedExperiment object:
cmap_expr <- saveHDF5SummarizedExperiment(cmap_expr, "./data/cmap_expr")

#########################
## make lincs database ##
#########################

# The `lincs` database represents moderate z-scores from differential expression 
# (DE) analysis of 12,328 genes from 8,140 compound treatments in 30 cells 
# (45,956 signatures in total). 

# Download the original LINCS datasets at GEO 
# <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742>
# The needed files are:
#   + GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx.gz
#   + GSE92742_Broad_LINCS_gene_info.txt.gz
#   + GSE92742_Broad_LINCS_sig_info.txt.gz
#   + GSE92742_Broad_LINCS_cell_info.txt.gz	
# Unzip the downloaded files to the `data` direcotry under the present working 
# directory of the user’s R session.

# The following codes require large number of memory in the system, 50Gb works
# Install cmapR package at <https://github.com/cmap/cmapR>
library(readr); library(dplyr); library(magrittr); library(cmapR)
library(SummarizedExperiment); library(HDF5Array)
meta42 <- read_tsv("./data/GSE92742_Broad_LINCS_sig_info.txt") 
txt <- names(table(meta42$pert_idose))
unit <- txt[grep("10 .M", txt)][2]
meta42 %<>% filter(pert_type=="trt_cp" & pert_idose==unit & pert_itime=="24 h") 
# filter rows by cmp, concentration, and treatment time
meta42 %<>% 
  bind_cols(alt_id=paste(meta42$pert_iname, meta42$cell_id, sep="_")) %>%
  bind_cols(pert_cell_factor=paste(meta42$pert_iname, meta42$cell_id, 
                                   meta42$pert_type, sep="__")) %>% 
  distinct(alt_id, .keep_all=TRUE) # eliminate technical duplicates
dim(meta42) # 45956 X 14
lincs42_mat <- cmapR::parse.gctx(
  "./data/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx", 
  cid=meta42$sig_id, matrix_only=TRUE)
lincs42_mat <- lincs42_mat@mat
colnames(lincs42_mat) = meta42$pert_cell_factor 
geneInfo <- read_tsv("./data/GSE92742_Broad_LINCS_gene_info.txt")
geneInfo <- as(geneInfo, "DataFrame")
geneInfo$pr_gene_id <- as.character(geneInfo$pr_gene_id)
rownames(geneInfo) <- geneInfo$pr_gene_id
geneInfo <- geneInfo[rownames(lincs42_mat),]
meta42 <- as(meta42, "DataFrame")
rownames(meta42) <- meta42$pert_cell_factor
lincs42 <- SummarizedExperiment(assays = list(score = lincs42_mat), 
                                rowData = geneInfo, colData = meta42)
# Get cell type information
cell_info <- read_tsv("./data/GSE92742_Broad_LINCS_cell_info.txt")
cell_name <- unique(meta42$cell_id)
cell_info %<>% dplyr::filter(base_cell_id %in% cell_name & !duplicated(
  paste(cell_info$base_cell_id, cell_info$sample_type, sep="_"))) %>%
  dplyr::select(c("cell_id","sample_type","primary_site","subtype")) %>% 
  dplyr::rename(cell_type=sample_type)
## replace "primary" as "normal" in cell_type column
cell_info %<>% mutate(cell_type=gsub("primary","normal",cell_type))
# Add cell info to lincs42 database
metadata(lincs42)$cell_info = cell_info
# Save 'lincs42' as an HDF5-based SummarizedExperiment object:
lincs42 <- saveHDF5SummarizedExperiment(lincs42, "./data/lincs")

##############################
## make lincs_expr database ##
##############################

# The `lincs_expr` database contains gene expression intensity values 
# from 5,925 compound treatments in 30 cells (38,824 signatures in total). 

# Download the original LINCS expression datasets at GEO 
# <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742>
# The needed files are:
#   + GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx.gz
#   + GSE92742_Broad_LINCS_gene_info.txt.gz
#   + GSE92742_Broad_LINCS_inst_info.txt.gz
# Unzip the downloaded files to the `data` direcotry under the present working 
# directory of the user’s R session.

# The following codes require large number of memory in the system, 50Gb works
inst42 <- read_tsv("./data/GSE92742_Broad_LINCS_inst_info.txt") 
inst42 %<>% filter(pert_type=="trt_cp" & pert_dose==10 & pert_dose_unit=="um" 
                   & pert_time==24 & pert_time_unit=="h")
inst42 %<>% 
  bind_cols(alt_id=paste(inst42$pert_iname, inst42$cell_id, sep="_")) %>% 
  bind_cols(pert_cell_factor=paste(inst42$pert_iname, inst42$cell_id, 
                                   inst42$pert_type, sep="__"))
dim(inst42) # 169,795 X 13
lincs42_mat3 <- cmapR::parse.gctx(
  "./data/GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx", 
  cid=inst42$inst_id, matrix_only=TRUE)
lincs42_mat3 <- lincs42_mat3@mat # 12328 X 169795
pert_cell_list <- split(inst42$inst_id, inst42$pert_cell_factor)
lincs_drug_cell_expr <- sapply(names(pert_cell_list), function(x) 
  rowMeans(as.data.frame(lincs42_mat3[,pert_cell_list[[x]]]))) # 12328 X 38824
saveRDS(lincs_drug_cell_expr, "./data/lincs_drug_cell_expr.rds")
geneInfo <- read_tsv("./data/GSE92742_Broad_LINCS_gene_info.txt")
geneInfo <- as(geneInfo, "DataFrame")
geneInfo$pr_gene_id <- as.character(geneInfo$pr_gene_id)
rownames(geneInfo) <- geneInfo$pr_gene_id
geneInfo <- geneInfo[rownames(lincs_drug_cell_expr),]
inst42_uniq <- inst42[!duplicated(inst42$pert_cell_factor),] # 38824 X 13
inst42_uniq <- as(inst42_uniq, "DataFrame")
rownames(inst42_uniq) <- inst42_uniq$pert_cell_factor
lincs42_expr <- SummarizedExperiment(
  assays = list(expr = lincs_drug_cell_expr), 
  rowData = geneInfo, 
  colData = inst42_uniq[colnames(lincs_drug_cell_expr),])
## Save 'lincs42_expr' as an HDF5-based SummarizedExperiment object:
lincs42_expr <- saveHDF5SummarizedExperiment(lincs42_expr, "./data/lincs_expr")

#################################
## make dtlink_db_clue_sti.db ##
#################################

# This SQLite database contains drug-target links (dtlink) in DrugBank, CLUE, 
# and STITCH databases

## Get dtlink in DrugBank database (version 5.1.2)

### Download the `drugbank_all_target_uniprot_links.csv.zip` at 
### <https://www.drugbank.ca/releases/latest#external-links> under 
### `Target Drug-UniProt Links` section where `DRUG GROUP` is `All`.
### Extract the downloaded zip file, rename the extracted csv file as 
### `drug_target_uniprot_links_5.1.2.csv`, save it under `data` directory 

dt_uni_link <- read.delim("./data/drug_target_uniprot_links_5.1.2.csv",
                        sep=",", stringsAsFactors=FALSE)
#### Add gn_sym column for dt_uni_link
library(org.Hs.eg.db)
k <- keys(org.Hs.eg.db, keytype = "UNIPROT")
uni_gnsym <- suppressMessages(AnnotationDbi::select(
  org.Hs.eg.db, keys=k, columns=c("UNIPROT", "SYMBOL"), keytype="UNIPROT"))
dt_uni_link <- dplyr::as_tibble(dt_uni_link)
dt_uni_sym <- dplyr::left_join(dt_uni_link[,c("Name","UniProt.ID")], 
                               uni_gnsym, by = c("UniProt.ID"="UNIPROT"))
dtlink_db <- na.omit(unique(data.frame(drug_name=tolower(dt_uni_sym$Name), 
                        t_gn_sym=dt_uni_sym$SYMBOL, stringsAsFactors=FALSE)))
#### exclude drugs that have more than 100 targets in dtlink_db
dtlist_db <- split(dtlink_db$t_gn_sym, dtlink_db$drug_name)
dtnum_db <- sapply(dtlist_db, length)
sum(dtnum_db>100)
dtlist_db <- dtlist_db[dtnum_db<=100]
dtlink_db <- list2link(dtlist_db, type="dt")
#### exclude proteins that are targeted by more than 100 drugs in dtlink_db
tdlist_db <- split(dtlink_db$drug_name, dtlink_db$t_gn_sym)
tdnum_db <- sapply(tdlist_db, length)
sum(tdnum_db>100)
tdlist_db <- tdlist_db[tdnum_db<=100]
tdlink_db <- list2link(tdlist_db, type="td")
dtlink_db <- tdlink_db[,c("drug_name","t_gn_sym")]
length(unique(dtlink_db$drug_name)) # 5250
length(unique(dtlink_db$t_gn_sym)) # 2443
write.table(dtlink_db, "./data/dtlink_db_5.1.2.xls", 
            quote=FALSE, col.names = NA, sep="\t")

## Get dtlink in CLUE touchstone database at <https://clue.io/touchstone>
### Get pertabation information in CLUE touchstone (version 1.1.1.2) from API

#### Query the CLUE website at <https://clue.io/query> to get drugs in 
#### Touchstone database:
#### At Query window, select `Gene expression (L1000)` as query parameters. 
#### Select `individual query` as query mode. Then click the `example` link and
#### choose `PI3K/MTOR inhibitor` as an example query, the name of your query,
#### UP-regulated genes and DOWN-regulated genes would be automatically filled.
#### Then click `SUBMIT` button.
#### At the Query History window, select the query result, then click the 
#### `DETAILED LIST` button
#### At the result window, select `Compound x 2429` under `Perturbagen Type`.
#### Click `Export` button to save the result as `clue_example_mtor_res.txt` 
#### to your data directory under current R session. 

clue_res <- read.delim("./data/clue_example_mtor_res.txt",sep="\t") # 2429 X 6
clue_drugs <- as.character(unique(clue_res$Name)) # 2397
clue_pert_info_api <- getPertInfo(lincs_drugs)
saveRDS(clue_pert_info_api, "./data/clue_pert_info_api.rds")
clue_dt <- clue_pert_info_api[,c("pert_iname","target")]
clue_dtlist <- sapply(split(clue_dt$target, clue_dt$pert_iname), 
                       function(x) unique(unlist(x))) # 2397
dtlink_clue <- list2link(clue_dtlist, type="dt")
dtlink_clue$drug_name <- tolower(dtlink_clue$drug_name)
#### exclude drugs that have more than 100 targets, 
#### then exclude genes that have more than 100 targeted drugs
dtlist_clue <- split(dtlink_clue$t_gn_sym, dtlink_clue$drug_name)
dtnum_clue <- sapply(dtlist_clue, length)
sum(dtnum_clue > 100) # 0

tdlist_clue <- split(dtlink_clue$drug_name, dtlink_clue$t_gn_sym)
tdnum_clue <- sapply(tdlist_clue, length)
sum(tdnum_clue > 100)
tdlist_clue <- tdlist_clue[tdnum_clue <= 100]
tdlink_clue <- list2link(tdlist_clue, type="td")
dtlink_clue <- tdlink_clue[,c("drug_name","t_gn_sym")]
length(unique(dtlink_clue$drug_name)) # 1949
length(unique(dtlink_clue$t_gn_sym)) # 1325
write.table(dtlink_clue, "data/dtlink_clue_1.1.1.2.xls", 
            quote=FALSE, sep="\t", row.names=FALSE)
dtlink_clue <- read.delim("./data/dtlink_clue_1.1.1.2.xls", stringsAsFactors=FALSE)

## Get dtlink in STITCH database (v5.0) 
### At STITCH Download page 
### <http://stitch.embl.de/cgi/download.pl?UserId=Z159fH9RYlUj&sessionId=2yGVo2kYC0ri>,
### Choose `Homo sapiens` organism, download and decompress the 
### `9606.protein_chemical.links.detailed.v5.0.tsv.gz` and
### `chemicals.v5.0.tsv.gz` as 
### `9606_protein_chemical_links_detailed_v5.0.tsv` and
### `chemicals.v5.0.tsv`, respectivily, to the data directory 
### under your current working directory of R session. 

### Need to allocate large number of memory to read the table, 20Gb works
sti_dtlink_detail <- readr::read_tsv(
  "data/9606_protein_chemical_links_detailed_v5.0.tsv") # 15,473,939 X 7
sti_dtlink <- dplyr::filter(sti_dtlink_detail, 
                            experimental > 700 | database > 700) # 346,909 X 7
### Eliminate drugs that have more than 100 targets in `sti_dtlink`
sti_dtlist <- split(sti_dtlink$protein, sti_dtlink$chemical)
sti_dtnum <- sapply(sti_dtlist, length)
invalid_drugs <- names(sti_dtlist[sti_dtnum>100]) # 356
sti_dtlink <- dplyr::filter(sti_dtlink, ! chemical %in% invalid_drugs)
sti_dtlink$protein <- sub("9606.", "", sti_dtlink$protein) # 227,948 X 7
### get PubChem CID to drug name mappings from `chemicals.v5.0.tsv` file
sti_chem_info <- readr::read_tsv("data/chemicals.v5.0.tsv")
### Map Ensembl protein ids to gene symbol 
library(EnsDb.Hsapiens.v75)
edb <- EnsDb.Hsapiens.v75
gnames <- keys(edb, keytype = "GENENAME")
ensb_gn_protid <- na.omit(select(edb, keys = gnames, keytype = "GENENAME", 
                                 columns = c("GENENAME", "PROTEINID")))
sti_dtlink %<>% left_join(ensb_gn_protid, by = c("protein" = "PROTEINID")) %>% 
    left_join(sti_chem_info[,c("chemical","name")]) %>% 
    dplyr::select(drug_name=name, t_gn_sym=GENENAME) # 227,948 X 2
sti_dtlink <- unique(na.omit(sti_dtlink)) # 142291 X 2
sti_dtlink$drug_name <- sapply(sti_dtlink$drug_name, 
                               function(x) unlist(strsplit(x,", "))[1])
sti_dtlink$drug_name <- tolower(sti_dtlink$drug_name)
### Exclude proteins that are targeted by more than 100 drugs in sti_dtlink
#### check whether there are drugs that have more than 100 targets 
#### after ENSP id to gene SYMBOL mappings
dtlist_sti <- split(sti_dtlink$t_gn_sym, sti_dtlink$drug_name)
dtnum_sti <- sapply(dtlist_sti, length)
sum(dtnum_sti>100) # 0
tdlist_sti <- split(sti_dtlink$drug_name, sti_dtlink$t_gn_sym)
tdnum_sti <- sapply(tdlist_sti, length)
sum(tdnum_sti>100) # 255
tdlist_sti <- tdlist_sti[tdnum_sti<=100]
tdlink_sti <- list2link(tdlist_sti, type="td")
dtlink_sti <- tdlink_sti[,c("drug_name","t_gn_sym")] # 50,012 X 2
length(unique(dtlink_sti$drug_name)) # 18,847
length(unique(dtlink_sti$t_gn_sym)) # 5,119
write.table(dtlink_sti, "./data/dtlink_sti_v5.0.xls", 
            quote=FALSE, sep="\t", row.names=FALSE)

#####################
## make goAnno.rds ##
#####################

# It is an intermediate file storing the GO to GENE SYMBOLE mapping information
# for `dtnetplot` function in `signatureSearch` package

OrgDb <- GOSemSim::load_OrgDb(OrgDb = "org.Hs.eg.db")
kk <- keys(OrgDb, keytype="SYMBOL")
goAnno <- suppressMessages(AnnotationDbi::select(OrgDb, keys=kk, keytype="SYMBOL",
                        columns=c("GOALL", "ONTOLOGYALL")))
goAnno <- unique(na.omit(goAnno)) # 3,430,396 X 4
saveRDS(goAnno, "./data/goAnno.rds")

##########################
## make goAnno_drug.rds ##
##########################

# It is an intermediate file storing the GO to drug name mapping information
# for `dsea` functions in `signatureSearch` package

## convert GOALL-SYMBOL in goAnno to GOALL-drug
## Get drug-target annotation in DrugBank, CLUE and STITCH databases from
## `dtlink_db_clue_sti.db` created above
library(RSQLite)
##
goAnno_drug <- left_join(as_tibble(goAnno[,c(2,1,4)]), as_tibble(dtlink), by = c("SYMBOL"="t_gn_sym"))
rm(dtlink)
goAnno_drug <- as.data.frame(goAnno_drug)[,c("GOALL","ONTOLOGYALL","drug_name")]
message("remove na in goAnno_drug")
goAnno_drug <- na.omit(goAnno_drug)
message("unique GOALL and drug_name in goAnno_drug")
goAnno_drug <- goAnno_drug[!duplicated(goAnno_drug[,c("GOALL","drug_name")]),]
message("unique is done")
# goAnno_drug <- goAnno_drug %>% filter(!is.na(drug_name)) %>% distinct(GOALL, drug_name, .keep_all = TRUE)
# goAnno_drug <- as.data.frame(goAnno_drug)

